runtime/doc/lua-plugin.txt

```diff
@@ -0,0 +1,447 @@
+
+                            NVIM REFERENCE MANUAL
+
+		  Guide to developing Lua plugins for Nvim
+
+
+                                       Type |gO| to see the table of contents.
+
+==============================================================================
+Introduction                                                       *lua-plugin*
+
+This is a guide for getting started with Neovim plugin development. It is not
+intended as a set of rules, but as a collection of recommendations for good
+practices.
+
+For a guide to using Lua in Neovim, please refer to |lua-guide|.
+
+==============================================================================
+Type safety                                            *lua-plugin-type-safety*
+
+Lua, as a dynamically typed language, is great for configuration. It provides
+virtually immediate feedback.
+But for larger projects, this can be a double-edged sword, leaving your plugin
+susceptible to unexpected bugs at the wrong time.
+
+You can leverage LuaCATS https://luals.github.io/wiki/annotations/    *LuaCATS*
+annotations, along with lua-language-server https://luals.github.io/ to catch
+potential bugs in your CI before your plugin's users do.
+
+------------------------------------------------------------------------------
+Tools						 *lua-plugin-type-safety-tools*
+
+- lua-typecheck-action https://github.com/marketplace/actions/lua-typecheck-action
+- llscheck https://github.com/jeffzi/llscheck
+- luacheck https://github.com/mpeterv/luacheck for additional linting
+- neodev.nvim https://github.com/folke/neodev.nvim
+
+==============================================================================
+User commands					     *lua-plugin-user-commands*
+
+Many users rely on command completion to discover available user commands. If
+a plugin pollutes the command namespace with lots of commands, this can
+quickly become overwhelming.
+
+Example:
+
+- `FooAction1 {arg}`
+- `FooAction2 {arg}`
+- `FooAction3`
+- `BarAction1`
+- `BarAction2`
+
+Instead of doing this, consider gathering subcommands under scoped commands
+and implementing completions for each subcommand.
+
+Example:
+
+- `Foo action1 {arg}`
+- `Foo action2 {arg}`
+- `Foo action3`
+- `Bar action1`
+- `Bar action2`
+
+------------------------------------------------------------------------------
+Subcommand completions example   *lua-plugin-user-commands-completions-example*
+
+In this example, we want to provide:
+
+- Subcommand completions if the user has typed `:Foo ...`
+- Argument completions if they have typed `:Foo {subcommand}`
+
+First, define a type for each subcommand, which has:
+
+- An implementation: A function which is called when executing the subcommand.
+- An optional command completion callback, which takes the lead of the
+  subcommand's arguments.
+
+>lua
+---@class FooSubcommand
+---@field impl fun(args:string[], opts: table)
+---@field complete? fun(subcmd_arg_lead: string): string[]
```

Comment:
You need to indent all codeblocks for TS to highlight correctly. Don't ask me why, thems the rules.

```suggestion
    ---@class FooSubcommand
    ---@field impl fun(args:string[], opts: table)
    ---@field complete? fun(subcmd_arg_lead: string): string[]
```

Reply:
Ah, I was wondering about that :smile: 
Good to know.